app/features/ai_resistant_assignment_generator/tools.py

```diff
@@ -0,0 +1,172 @@
+import os
+import json
+import time
+import requests
+import pandas as pd
+from langchain_core.prompts import PromptTemplate
+# from langchain_google_vertexai import VertexAI
+import re
+# from reportlab.lib.pagesizes import letter
+# from reportlab.pdfgen import canvas
+# from reportlab.platypus import Table, TableStyle
+# from reportlab.lib import colors
+from docx import Document
+from requests.exceptions import HTTPError
+from io import BytesIO
+# from bs4 import BeautifulSoup
+import requests
+# import praw
+# import prawcore
+from app.services.logger import setup_logger
+from langchain_groq import ChatGroq
+import requests
+import os
+from io import BytesIO
+from PyPDF2 import PdfReader
+from docx import Document
+#from app.features.syllabus_generator import credentials
+
+logger = setup_logger(__name__)
+
+model_name = 'llama-3.1-70b-versatile'
+# Entire syllabus generator pipeline with all functions in this class
+class AIRAG :
+
+    def __init__(self,grade,assignment,description='',path=""):
+        self.grade = grade
+        self.assignment = assignment
+        self.description = description
+        self.path = path
+        self.model = ChatGroq(model=model_name,temperature=0.3,api_key="")
+
+    def read_text_file(self,filepath):
+
+        with open(f"{self.path}{filepath}", 'r') as file:
+            return file.read()
+
+    def build_prompt(self,filepath):
+        # build invididual promps for each sub part of syllabus
+        template = self.read_text_file(filepath)
+
+        prompt = PromptTemplate.from_template(template)
+        return prompt
+
+    def validator(self,response):
+        data = []
+        try:
+            data = json.loads(response)
+        except json.JSONDecodeError as e:
+
+           
+            print("JSON Decode Error , Trying to correct the JSON")
+            try:
+                corrected_result = response[min(response.find('{'),response.find('[')):max(response.rfind(']'),response.rfind('}')) + 1]
+                data = json.loads(corrected_result)
+                print("Corrected Parsed JSON successfully")
+            except json.JSONDecodeError as e:
+                print("Failed to parse corrected JSON")
+        return data
+    
+    def extract_text_from_txt(self,file_content):
```

Comment:
You can create a reusable function for your document loaders and use the built-in document loaders from LangChain for a better enhancement of the documents extraction process.